### PR TITLE
PMM-13078 List available client versions

### PR DESCRIPTION
**Pull request by ademidoff**
PMM-13078

Link to the Feature Build: SUBMODULES-3666

### CODE REVIEW

File: managed/services/management/agent.go [hunk 1]
```diff
@@ -207,3 +210,79 @@ func (s *ManagementService) validateListAgentRequest(req *managementv1.ListAgent
 
 	return nil
 }
+
+func (s *ManagementService) ListAgentVersions(ctx context.Context, _ *managementv1.ListAgentVersionsRequest) (*managementv1.ListAgentVersionsResponse, error) {
```
[the remark this answers is not in the document]
Comment (ademidoff) [reply managed/services/management/agent.go:214]: Fixed.

File: api/management/v1/agent.proto [hunk 2]
```diff
@@ -146,3 +146,31 @@ message ListAgentsResponse {
...
+  string node_name = 3;
+  // Update severity.
+  UpdateSeverity severity = 4;
+}
+
+message ListAgentVersionsRequest {}
+
+message ListAgentVersionsResponse {
+  // List of Agent versions.
+  repeated AgentVersions versions = 1;
```
[the remark this answers is not in the document]
Comment (ademidoff) [reply api/management/v1/agent.proto]: Done.

File: managed/services/management/agent.go [hunk 3]
```diff
@@ -207,3 +210,80 @@ func (s *ManagementService) validateListAgentRequest(req *managementv1.ListAgent
...
+			}
+
+			var severity managementv1.UpdateSeverity
+			switch {
+			case agentVersion.Major < serverVersion.Major:
+				severity = managementv1.UpdateSeverity_UPDATE_SEVERITY_CRITICAL
+			case agentVersion.Less(serverVersion):
+				severity = managementv1.UpdateSeverity_UPDATE_SEVERITY_REQUIRED
+			case serverVersion.Less(agentVersion):
+				return status.Errorf(codes.Internal, "unsupported agent version: %s", *agent.Version)
```
[the remark this answers is not in the document]
Comment (ademidoff) [reply managed/services/management/agent.go]: Sure thing! I added one more severity level for unsupported clients.

File: Dockerfile [hunk 4]
```diff
@@ -8,4 +8,5 @@ RUN export GOPATH=$(go env GOPATH) && \
 COPY . $GOPATH/src/github.com/percona/pmm/
 WORKDIR $GOPATH/src/github.com/percona/pmm/api-tests/
 
-CMD make init run-race
+CMD ["make", "init", "run-race"]
```
Comment (ademidoff) [inline Dockerfile:11]: This allows to get rid of the docker build warning that says that it's better to use `exec` syntax vs `bash` syntax for RUN, CMD and ENTRYPOINT directives.

**Review by BupycHuk — APPROVED** [review]

**Review by idoqo — APPROVED** [review]

File: managed/services/management/agent.go [hunk 5]
```diff
@@ -207,3 +210,80 @@ func (s *ManagementService) validateListAgentRequest(req *managementv1.ListAgent
 
 	return nil
 }
+
+// ListAgentVersions returns a list of agents with their update recommendations (updte severity).
```
[the remark this answers is not in the document]
Comment (ademidoff) [reply managed/services/management/agent.go:214]: Done.

File: managed/services/management/agent.go [hunk 6]
```diff
@@ -207,3 +210,80 @@ func (s *ManagementService) validateListAgentRequest(req *managementv1.ListAgent
...
+			nodeName, ok := nodeNames[pointer.GetString(agent.RunsOnNodeID)]
+			if !ok {
+				s.l.Warnf("node not found for agent %s", agent.AgentID)
+				continue
+			}
+
+			agentVersion, err := version.Parse(pointer.GetString(agent.Version))
+			if err != nil {
+				// We don't want to fail the whole request if we can't parse the agent version.
+				s.l.Warnf(errors.Wrap(err, fmt.Sprintf("could not parse the client version: %s", pointer.GetString(agent.Version))).Error())
```
[the remark this answers is not in the document]
Comment (ademidoff) [reply managed/services/management/agent.go:255]: Good point, done.

**Pull request merged**